src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -328,10 +395,42 @@ uct_dc_mlx5_ep_basic_init(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
                           UCS_STATS_ARG(ep->super.stats));
 }
 
-static UCS_F_ALWAYS_INLINE int
+static int
 uct_dc_mlx5_iface_dci_can_alloc(uct_dc_mlx5_iface_t *iface, uint8_t pool_index)
 {
-    return iface->tx.dci_pool[pool_index].stack_top < iface->tx.ndci;
+    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
+
+    if (ucs_likely(pool->stack_top < ucs_array_length(&pool->stack))) {
+        return 1;
+    }
+
+    if ((ucs_array_length(&pool->stack) == iface->tx.ndci) ||
+        (pool->stack_top == iface->tx.ndci)) {
+        return 0;
+    }
+
+    return 1;
```

Comment:
return (ucs_likely(pool->stack_top < ucs_array_length(&pool->stack))) || 
((ucs_array_length(&pool->stack) < iface->tx.ndci) && (pool->stack_top != iface->tx.ndci)
BTW, why needed to check pool->stack_top? it wasn't checked in previous code

Reply:
**can remove 408 and unify conditions

Reply:
can_alloc -> revert to only check stack_top
can_alloc_or_create -> check stack_top and the rest of the cases